homeassistant/components/sensor/ztm.py

```diff
@@ -0,0 +1,198 @@
+            'line': line,
+        }
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._unit
+
+    @property
+    def device_state_attributes(self):
+        """Return extra attributes."""
+        attribution = CONF_ATTRIBUTION
+        if self._timetable_date:
+            attribution_date = " on {}".format(self._timetable_date)
+            attribution = attribution + attribution_date
+        self._attributes[ATTR_ATTRIBUTION] = attribution
+        return self._attributes
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Update state."""
+        if self.data_is_outdated():
+            res = yield from async_http_request(self._loop, self._websession,
+                                                self._uri, self._params)
+            if res.get('error', ''):
+                self._state = "Error: {}".format(res['error'])
+                self._unit = ''
+            else:
+                self._timetable = self.map_results(res.get('results', []))
+                self._timetable_date = dt_util.now().date()
+                _LOGGER.info("Downloaded timetable for line:%s stop:%s-%s",
+                             self._line, self._stop_id, self._stop_number)
+
+        # check if there are trains after actual time
+        departures = []
+        now = dt_util.now()
+        for entry in self._timetable:
+            entry_time = dt_util.parse_time(entry['czas'])
+            entry_dt = datetime.combine(now.date(), entry_time)
+            entry_dt = entry_dt.replace(tzinfo=now.tzinfo)
+            if entry_dt > now:
+                time_left = int((entry_dt - now).seconds / 60)
+                departures.append(time_left)
+                if len(departures) == self._entries:
+                    break
+        if departures:
+            self._state = departures[0]
+            self._attributes['departures'] = departures
+            self._unit = 'min'
+        else:
+            self._state = STATE_UNKNOWN
+            self._unit = ''
+
+    def data_is_outdated(self):
+        """Check if the internal sensor data is outdated."""
+        now = dt_util.now()
+        return self._timetable_date != now.date()
+
+    @staticmethod
+    def map_results(response):
+        """Map all timetable entries to proper {'key': 'value'} struct."""
+        return [parse_raw_timetable(row) for row in response['result']]
+
+
+def parse_raw_timetable(raw_result):
+    """Change {'key': 'name','value': 'val'} into {'name': 'val'}."""
+    result = {}
+    for val in raw_result.get('values', []):
```

Comment:
Using an empty list as default value if the key is missing, will lead to error below.